calliope/test/test_backend_pyomo_interface.py

```diff
@@ -216,3 +216,118 @@ def test_rerun_fail_on_operate(self, model):
+            )
+
+        constraint_name = "energy_cap_time_varying"
+        constraint_sets = ["loc_techs_finite_resource", "timesteps"]
+        with pytest.raises(AssertionError) as excinfo:
+            model.backend.add_constraint(
+                constraint_name, constraint_sets, energy_cap_time_varying_rule
+            )
+        assert check_error_or_warning(
+            excinfo, "First argument of constraint function must be 'backend_model'."
+        )
+
+    def test_arg_mismatch(self, model):
+        """length of function arguments = length of sets + 1"""
+
+        def energy_cap_time_varying_rule(backend_model, loc_tech, timestep, extra_arg):
+
+            return (
+                backend_model.energy_cap[loc_tech]
+                <= backend_model.energy_cap[loc_tech]
+                * backend_model.resource[loc_tech, timestep]
+                + extra_arg
+            )
+
+        constraint_name = "energy_cap_time_varying"
+        constraint_sets = ["loc_techs_finite_resource", "timesteps"]
+        with pytest.raises(AssertionError) as excinfo:
+            model.backend.add_constraint(
+                constraint_name, constraint_sets, energy_cap_time_varying_rule
+            )
+        assert check_error_or_warning(
+            excinfo,
+            "Number of constraint arguments must equal number of constraint sets + 1.",
+        )
+
+    def test_sets(self, model):
+        """Constraint sets must be backend model sets"""
+
+        def energy_cap_time_varying_rule(backend_model, loc_tech, not_a_set):
+
+            return (
+                backend_model.energy_cap[loc_tech]
+                <= backend_model.energy_cap[loc_tech]
+                * backend_model.resource[loc_tech, not_a_set]
+            )
+
+        constraint_name = "energy_cap_time_varying"
+        constraint_sets = ["loc_techs_finite_resource", "not_a_set"]
+        with pytest.raises(AttributeError) as excinfo:
+            model.backend.add_constraint(
+                constraint_name, constraint_sets, energy_cap_time_varying_rule
+            )
+
+        assert check_error_or_warning(
+            excinfo, "Pyomo backend model object has no attribute 'not_a_set'"
+        )
+
+    def test_added_constraint(self, model):
+        """Constraint sets must be backend model sets"""
+
+        def new_constraint_rule(backend_model, loc_tech_carrier, timestep):
+            loc_tech = calliope.backend.pyomo.util.get_loc_tech(loc_tech_carrier)
+            carrier_con = backend_model.carrier_con[loc_tech_carrier, timestep]
+            timestep_resolution = backend_model.timestep_resolution[timestep]
+            return carrier_con * 2 >= (
+                -1 * backend_model.energy_cap[loc_tech] * timestep_resolution
+            )
+
+        constraint_name = "new_constraint"
+        constraint_sets = [
+            "loc_tech_carriers_carrier_consumption_max_constraint",
```

Comment:
As in the documentation, I don't really understand what this set is. Isn't there a `loc_tech_carrier` set that you could use here?

Reply:
that is a `loc_tech_carrier` set, but I'll change it to `loc_tech_carriers_con` for simplicity (I *think* they're equivalent in this instance)